src/NuGet.Clients/NuGet.Credentials/CredentialService.cs

```diff
@@ -51,13 +51,13 @@ public class CredentialService : ICredentialService
 
             ErrorDelegate = errorDelegate;
             _nonInteractive = nonInteractive;
-            Providers = providers ?? new List<ICredentialProvider>();
+            Providers = new Lazy<IEnumerable<ICredentialProvider>>(providersDelegate);
         }
 
         /// <summary>
         /// Gets the currently configured providers.
         /// </summary>
-        public IEnumerable<ICredentialProvider> Providers { get; } = null;
+        public Lazy<IEnumerable<ICredentialProvider>> Providers { get; } = null;
```

Comment:
To maintain the contract I'd keep the original property and would add a backup field, e.g. `_providersFactory` instead.